internal/server/server.go

```diff
@@ -1,24 +1,152 @@
 package server
 
 import (
-	"fmt"
+	"crypto/tls"
 	"net/http"
+	"net/http/httputil"
+	"net/url"
+
+	gapi "github.com/grafana/grafana-api-golang-client"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/config"
+	"github.com/kanopy-platform/grafana-auth-proxy/internal/jwt"
+	"github.com/kanopy-platform/grafana-auth-proxy/pkg/grafana"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	grafanaAuthHeader = "X-WEBAUTH-USER"
 )
 
 type Server struct {
-	router *http.ServeMux
+	router          *http.ServeMux
+	cookieName      string
+	groups          config.Groups
+	grafanaProxyUrl *url.URL
+	grafanaClient   *grafana.Client
+	skipTLSVerify   bool
 }
 
-func New() http.Handler {
+type ServerFuncOpt func(*Server) error
+
+func New(opts ...ServerFuncOpt) (http.Handler, error) {
 	s := &Server{router: http.NewServeMux()}
 
+	// load options
+	for _, opt := range opts {
+		err := opt(s)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	s.router.HandleFunc("/", s.handleRoot())
 
-	return s.router
+	return s.router, nil
 }
 
 func (s *Server) handleRoot() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		fmt.Fprintln(w, "hello world")
+		// Extract token from cookie
+		cookie, err := r.Cookie(s.cookieName)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading cookie")
+			return
+		}
+
+		// Get claims from token
+		claims, err := jwt.TokenClaims(cookie.Value)
+		if err != nil {
+			logAndError(w, http.StatusUnauthorized, err, "error reading claims from jwt token")
+			return
+		}
+
+		log.Infof("attempting to log in user %s", claims.Subject)
+		log.Debugf("claim groups for user %s: %v", claims.Subject, claims.Groups)
+
+		// validUserGroups represents the intersection of user groups from claim with the group
+		// mapping in configuration
+		validUserGroups := config.UserGroupsInConfig(claims.Groups, s.groups)
+		log.Debugf("valid user groups for user %s: %v", claims.Subject, validUserGroups)
+
+		// Creat gapi.User for lookup purposes
+		orgUser := gapi.User{
+			Login: claims.Subject,
+		}
+
+		// Mapping of role per org
+		userOrgsRole := make(map[int64]models.RoleType)
+
+		if claims.Email != "" {
```

Comment:
isn't the Email claim optional?

Reply:
Yes, the email claim is optional as per the OIDC spec.  The default value of the orgUser.Email is also an empty string.  So is their value in checking if the `claims.Email` contains a value? Since if not included it would also be an empty string.

If both are empty, isn't this `orgUser.Email = claims.Email` always valid?

Reply:
Do we care about pushing the email claim to the grafana global and org users? It doesn't seem like we do since I imagine it is used for maybe password reset which is hopefully disabled when this proxy is enabled.

Reply:
correct, it's not a hard requirement but I thought it would be useful if the information is available on the claims.

Reply:
applied David's suggestion

Reply:
I think if the auth proxy takes full control over how a user authenticates and user info, it should make sure the grafana user object accurately reflects the info in the jwt.

Reply:
agreed, it will basically populate as much as it can from the auth source, in this case the jwt

Reply:
We only upsert org users so if that is the case should this be an update on change to existing global users also?

Reply:
every user is an Org User by definition. Users are at least assigned to the default Org (1). In the code logic the user is always updated on the corresponding Org, even if that's Org 1, so the update path will always trigger and affect what is considered a "global user".